From 25b6f614d5da80037b98c3a174eab0b147f12095 Mon Sep 17 00:00:00 2001
From: Varun Sethi <Varun.Sethi@freescale.com>
Date: Thu, 18 Jun 2015 17:20:43 +0530
Subject: [PATCH 047/124] arm/smmu: PCIe controller stream ID partitioning
 patch.

U-boot sets up the device Id to stream ID translation table. The SMMU
driver while creating the IOMMU group programs the hardware table with
these values.

Signed-off-by: Varun Sethi <Varun.Sethi@freescale.com>
[Stuart: split the pci driver parts into a separate patch,
also updated to reflect new domain init/free ops]
Signed-off-by: Stuart Yoder <stuart.yoder@freescale.com>
---
 drivers/iommu/arm-smmu.c | 24 ++++++++++++++++++++++--
 1 file changed, 22 insertions(+), 2 deletions(-)

diff --git a/drivers/iommu/arm-smmu.c b/drivers/iommu/arm-smmu.c
index dd77129..40d3cb9 100644
--- a/drivers/iommu/arm-smmu.c
+++ b/drivers/iommu/arm-smmu.c
@@ -50,6 +50,10 @@
 #include <../drivers/staging/fsl-mc/include/mc.h>
 #endif
 
+#ifdef CONFIG_PCI_LAYERSCAPE
+#include <../drivers/pci/host/pci-layerscape.h>
+#endif
+
 #include <asm/pgalloc.h>
 
 /* Maximum number of stream IDs assigned to a single device */
@@ -1377,11 +1381,17 @@ static int arm_smmu_add_pci_device(struct pci_dev *pdev)
 	u16 sid;
 	struct iommu_group *group;
 	struct arm_smmu_master_cfg *cfg;
+#ifdef CONFIG_PCI_LAYERSCAPE
+       u32 streamid;
+#endif
 
 	group = iommu_group_get_for_dev(&pdev->dev);
 	if (IS_ERR(group))
 		return PTR_ERR(group);
 
+	if (pci_is_bridge(pdev))
+		goto out_put_group;
+
 	cfg = iommu_group_get_iommudata(group);
 	if (!cfg) {
 		cfg = kzalloc(sizeof(*cfg), GFP_KERNEL);
@@ -1412,6 +1422,16 @@ static int arm_smmu_add_pci_device(struct pci_dev *pdev)
 	if (i == cfg->num_streamids)
 		cfg->streamids[cfg->num_streamids++] = sid;
 
+#ifdef CONFIG_PCI_LAYERSCAPE
+	streamid = set_pcie_streamid_translation(pdev, sid);
+	if (~streamid == 0) {
+		ret = -ENODEV;
+		goto out_put_group;
+	}
+	cfg->streamids[0] = streamid;
+	cfg->mask = 0x7c00;
+#endif
+
 	return 0;
 out_put_group:
 	iommu_group_put(group);
@@ -1583,8 +1603,8 @@ static void arm_fsl_mc_smmu_remove_device(struct device *dev)
 
 static struct iommu_ops arm_fsl_mc_smmu_ops = {
 	.capable		= arm_smmu_capable,
-	.domain_init		= arm_smmu_domain_init,
-	.domain_destroy		= arm_smmu_domain_destroy,
+	.domain_alloc		= arm_smmu_domain_alloc,
+	.domain_free		= arm_smmu_domain_free,
 	.attach_dev		= arm_smmu_attach_dev,
 	.detach_dev		= arm_smmu_detach_dev,
 	.map			= arm_smmu_map,
-- 
2.1.0.27.g96db324

